oecophylla/taxonomy/taxonomy.rule

```diff
@@ -243,63 +250,63 @@ rule taxonomy_shogun:
                   --output {output.profile} \
                   2> {log} 1>&2
 
-                  # redistribute counts to certain taxonomic rank
-                  shogun redistribute \
-                  --database {params.db} \
-                  --level {params.level} \
-                  --input {output.profile} \
-                  --output {output.redist} \
-                  2> {log} 1>&2
+                  # keep mapping file
+                  if [[ "{params.map}" == "True" ]]
+                  then
+                    gzip -c {temp_dir}/alignment.{params.aligner}.{ext} > $stem.{params.aligner}.{ext}.gz
+                  fi
+
+                  # redistribute reads to given taxonomic ranks
+                  if [[ ! -z {params.levels} ]]
+                  then
+                    IFS=',' read -r -a levels <<< "{params.levels}"
+                    for level in "${{levels[@]}}"
```

Comment:
Can we simplify this?  It is not ideal to embed bash for loops inside of snakemake rules (its the problem that snakemake is trying to get away from).

Reply:
@mortonjt the issue is that there's a big imbalance in the compute times and requirements here. It takes a lot time/memory/disk to do the Kraken run, and the Bracken runs at each tax level are fast/cheap. So it's better to just do them all while the Kraken map output is in scratch disk rather than transferring back to filesystem and spinning up an extra n * levels jobs...

Reply:
Agree with @tanaes . Also noticed that multiple shell commands in one rule requires repeatedky activating and deactivating conda environments, which is inefficient. For example,
```
shell（"""kraken input.fq > output.txt""")
for level in levels:
    shell("""bracken output.txt > {level}.txt""")
```
This will not work, unless you `source activate` in each shell command.